### PR TITLE
handle automatic batch size in the training loop instead of up front

### PR DESCRIPTION
**Pull request by gnieuwenhuis**
The previous batch size calculation was failing once scaled to all coralnet sources (up to 2023), resulting in OOM failures in Sagemaker.

This reworks that logic by calculating batch size in the training loop after initialization is complete, and we no longer need to guess how much overhead memory was used for everything else, which avoids estimating a significant unknown in memory use calculations. Batch size is then estimated by making an estimate of memory needed only for training, which should be much more accurate.

Testing in Sagemaker has shown that I can train the classifier on all coralnet sources (up to 2023), where memory spikes up to 80%, but generally sits around 65% ro 70% during training. We could probably decrease the safety net from 20% to 10% if necessary in the future.

### CODE REVIEW

File: mermaid_classifier/pyspacer/settings.py [hunk 1]
```diff
@@ -5,37 +5,60 @@
...
-    rough_available_ram_bytes = max(total_ram_bytes - base_overhead_bytes, 0)
+    if clf_type == 'MLP':
+        # MLP hidden_layer_sizes is (200, 100) for large datasets or
+        # (100,) for small ones; num_classes is the output layer.
+        # Use the larger architecture as a conservative estimate.
+        activation_units = 200 + 100 + num_classes
+        activation_bytes = activation_units * _BYTES_PER_FLOAT
+    else:
+        # SGDClassifier has negligible internal buffers.
+        activation_bytes = 0
```
[the remark this answers is not in the document]
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/settings.py:55]: Applied. Doubled the MLP activation estimate (`2 * activation_units * _BYTES_PER_FLOAT`) to account for sklearn's per-layer backprop deltas and gradient buffers alongside the forward activations. This is in the conservative direction (smaller batch), matching the PR's goal of avoiding OOMs.

File: mermaid_classifier/pyspacer/settings.py [hunk 2]
```diff
@@ -81,9 +104,9 @@ class Settings(BaseSettings):
     mlflow_tracking_server: str | None = None
     training_inputs_percent_missing_allowed: int = 0
     spacer_extractors_cache_dir: str | None = None
-    # Yes, this is str. After it gets through the settings machinery,
-    # pyspacer will convert to int.
-    spacer_batch_size: str | None = str(automatic_batch_size())
+    # Override for training batch size. If None (default),
+    # training_batch_size() auto-calculates at runtime.
+    spacer_batch_size: str | None = None
```
[the remark this answers is not in the document]
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/settings.py]: Changed `spacer_batch_size` to `int | None`. Pydantic now validates the env var at startup — verified that a non-numeric value raises a ValidationError on `Settings()` rather than failing after data-prep, and removed the now-redundant `int(...)` cast in train.py.

File: mermaid_classifier/pyspacer/train.py [hunk 3]
```diff
@@ -1472,8 +1473,21 @@ def run(self, run_name: str | None = None):
             model_loc = DataLocation('memory', key='classifier.pkl')
             valresult_loc = DataLocation('memory', key='valresult.json')
 
-            batch_size = int(settings.spacer_batch_size)
-            logger.info(f"Batch size: {batch_size}")
+            clf_type = 'MLP'
```
Comment (yeelauren) [inline mermaid_classifier/pyspacer/train.py:1476]: will this ever be anything else than MLP ?

**Review by yeelauren — APPROVED** [review]
LGTM

File: mermaid_classifier/pyspacer/train.py (hunk 3 above)
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/train.py:1476]: No - I have another PR in the chain (https://github.com/data-mermaid/mermaid-classifier/pull/63) to remove this branching logic
